contracts/Authorize.sol

```diff
@@ -9,34 +9,62 @@ contract Authorize is Ownable, AccessControl {
   bytes32 public constant PATIENT_ROLE = keccak256("PATIENT");
   bytes32 public constant DOCTOR_ROLE = keccak256("DOCTOR");
 
+  mapping(address => address[]) private managerToDoctors;
+
   constructor() {
     _grantRole(DEFAULT_ADMIN_ROLE, _msgSender());
     _grantRole(MANAGER_ROLE, _msgSender());
+    _setRoleAdmin(DOCTOR_ROLE, MANAGER_ROLE);
+  }
+
+  modifier onlyManager(address _managerAddress) {
+    require(hasRole(MANAGER_ROLE, _managerAddress), "You are not authorized");
+    _;
   }
 
   // Only accounts with DEFAULT_ADMIN_ROLE role can set another account role to manager
-  function setManager(address userAddress) public {
-    grantRole(MANAGER_ROLE, userAddress);
+  function setManager(address _userAddress) public {
+    grantRole(MANAGER_ROLE, _userAddress);
   }
 
-  function setPatient(address userAddress) public {
-    _grantRole(PATIENT_ROLE, userAddress);
+  function setPatient(address _userAddress) public {
+    _grantRole(PATIENT_ROLE, _userAddress);
+  }
+
+  function setDoctor(address _managerAddress, address _userAddress) public onlyManager(_managerAddress) {
+    _grantRole(DOCTOR_ROLE, _userAddress);
+    if(!includes(managerToDoctors[_managerAddress], _userAddress)) {
+      managerToDoctors[_managerAddress].push(_userAddress);
+    }
   }
 
-  function getRole(address userAddress) external view returns (uint[] memory) {
+  function getDoctors(address _managerAddress) external view onlyManager(_managerAddress) returns (address[] memory) {
+    return managerToDoctors[_managerAddress];
+  }
+
+  function getRole(address _userAddress) external view returns (uint[] memory) {
     uint[] memory roleArray = new uint[](4);
-    if (hasRole(PATIENT_ROLE, userAddress)) {
+    if (hasRole(PATIENT_ROLE, _userAddress)) {
       roleArray[0] = 1;
     }
-    if (hasRole(DOCTOR_ROLE, userAddress)) {
+    if (hasRole(DOCTOR_ROLE, _userAddress)) {
       roleArray[1] = 1;
     }
-    if (hasRole(MANAGER_ROLE, userAddress)) {
+    if (hasRole(MANAGER_ROLE, _userAddress)) {
       roleArray[2] = 1;
     }
-    if (hasRole(DEFAULT_ADMIN_ROLE, userAddress)) {
+    if (hasRole(DEFAULT_ADMIN_ROLE, _userAddress)) {
       roleArray[3] = 1;
     }
     return roleArray;
   }
+
+  function includes(address[] memory _arr, address _element) private pure returns (bool) {
+    for (uint i = 0; i < _arr.length; i++) {
+      if(_element == _arr[i]) {
+        return true;
+      }
+    }
+    return false;
+  }
```

Comment:
@kkinggmann ở đoạn này thì t thấy là nếu hàm include này được gọi ở hàm setDoctor thì nó sẽ vẫn mất phí do hàm setDoctor với cả khi chạy vòng for sẽ mất nhiều phí gas thì tao đề xuất như này:
  mapping(address => address[]) private managerToDoctors;
  mapping(address => address) private doctorToManager;
nếu làm như này thì sẽ k cần hàm for mà chỉ cần doctorToManager[_userAddress] sẽ ra được thằng quản lý luôn

Reply:
bỏ được hàm này luôn